### ISSUE TITLE
[BUG] Tracer messages out of order

### ISSUE THREAD

**Issue by vladpaiu** [OpenSIPS/opensips#1891]
Hello,

Commit aaa6b6867763a564d5b53460ab3f23a4033272a6 introduced a bug related to the ordering of the tracing messages -  since the ACK callback gets called before TMCB_RESPONSE_IN, when tracing via HEP the output PCAPs are screwed, since the ACK is shown before the actual reply.

Regards,
Vlad

**Comment by bogdan-iancu** [comment]
naughty bot...

**Comment by bogdan-iancu** [comment]
@vladpaiu , I agree, the commit did change the order, but the tricky question is what should be the right approach ? 
If we trace the reply first, we may end up with delays at SIP level for sending back the ACK (so the SIP may be affected and delayed by the tracing operation).
If we trace the reply after sending back the ACK, we mess up with the tracing (ACK will be trace before the actual reply).
Part of the problem is that the ACK tracing is built into the sending function, you cannot send the ACK and trace it later (especially because the ACK buffer is ephemeral)

**Comment by nikbyte** [comment]
I expect the same issue. Please, don't close.

**Comment by gmaruzz** [comment]
+1

**Comment by bogdan-iancu** [comment]
Unfortunately there is no easy solution here, see my prev [comment](https://github.com/OpenSIPS/opensips/issues/1891#issuecomment-572189825)

**Comment by liviuchircu** [comment]
A [fix attempt](https://github.com/OpenSIPS/opensips/commit/43477da50c67eaa4b7bd206c2d8bcaa09b0b6b25) was pushed to `master` today, let's see if it holds and we can backport it down to `3.2` if there is positive feedback.
